test/adapters/show-document-adapter.test.ts

```diff
@@ -0,0 +1,22 @@
+import * as sinon from "sinon"
+import { expect } from "chai"
+import * as ShowDocumentAdapter from "../../lib/adapters/show-document-adapter"
+import { LanguageClientConnection } from "../../lib/languageclient"
+import { createSpyConnection } from "../helpers"
+
+describe("ShowDocumentAdapter", () => {
+  describe("can attach to a server", () => {
+    const lcc = new LanguageClientConnection(createSpyConnection())
+
+    it("subscribes to onShowDocument", async () => {
+      const spy = sinon.spy()
+      lcc["_onRequest"] = spy
+
+      ShowDocumentAdapter.attach(lcc)
+      expect((lcc["_onRequest"] as sinon.SinonSpy).calledOnce).to.be.true
```

Comment:
I feel like this is a good example of testing the implementation (which is a bad practice). We should be testing that the correct arguments are sent to the rpc connection.

Reply:
`attach` function is very simple. This is all we want. I will add tests for the actual `showDocument`

Reply:
But this still isn't testing the correct thing. `_onRequest` is a private method so if we ever decide to remove it and inline it's function wherever it is used we shouldn't have to change any tests since it is an implementation detail.

The correct test should be something like:

```
  describe("can attach to a server", () => {
    const connection = createSpyConnection()
    const lcc = new LanguageClientConnection(connection)

    it("subscribes to onShowDocument", async () => {
      const spy = sinon.spy()
			ShowDocumentAdapter.showDocument = spy
      ShowDocumentAdapter.attach(lcc)
      expect((connection.onRequest as sinon.SinonSpy).calledOnce).to.be.true
      const spyArgs = connection.onRequest.firstCall.args
      expect(spyArgs[0]).to.deep.equal({ method: "window/showDocument" })
			spyArgs[1]("test")
      expect((ShowDocumentAdapter.showDocument as sinon.SinonSpy).calledOnce).to.be.true
      expect(ShowDocumentAdapter.showDocument.firstCall.args).to.equal("test")
    })
  })
})
```

Reply:
Or we would at least have to test that `_onRequest` actually does what it is supposed to as well. Right now we are only testing the attach function half way.

Reply:
Fixed in https://github.com/atom-community/atom-languageclient/pull/135/commits/43e195dfd551278a2970c19bf6266a9b4cf74fb5

Yes, we can add tests for `_onRequest`. Because this PR doesn't edit that, the test for it is out of the scope of this PR.